test/unit/local/steps/dispatch.js

```diff
@@ -0,0 +1,389 @@
+/* eslint-env mocha */
+/* @flow */
+
+const should = require('should')
+const sinon = require('sinon')
+
+const Builders = require('../../../support/builders')
+const configHelpers = require('../../../support/helpers/config')
+const pouchHelpers = require('../../../support/helpers/pouch')
+
+const SyncState = require('../../../../core/syncstate')
+const Prep = require('../../../../core/prep')
+const Buffer = require('../../../../core/local/steps/buffer')
+const dispatch = require('../../../../core/local/steps/dispatch')
+
+describe('dispatch', function () {
+  let builders
+  let buffer
+  let events
+  let prep
+
+  before('instanciate config', configHelpers.createConfig)
+  beforeEach('instanciate pouch', pouchHelpers.createDatabase)
+  beforeEach('populate pouch with documents', function () {
+    builders = new Builders({pouch: this.pouch})
+    buffer = new Buffer()
+
+    events = sinon.createStubInstance(SyncState)
+    prep = sinon.createStubInstance(Prep)
+  })
+  afterEach('clean pouch', pouchHelpers.cleanDatabase)
+  after('clean config directory', configHelpers.cleanConfig)
+
+  context('when buffer contains an initial-scan-done event', () => {
+    beforeEach(() => {
+      buffer.push([
+        builders.event().action('initial-scan-done').build()
+      ])
+    })
+
+    it('emits an initial-scan-done event via the emitter', async function () {
+      const dispatchedBuffer = dispatch(buffer, { events, prep, pouch: this.pouch })
+      await dispatchedBuffer.pop()
```

Comment:
We're popping the dispatched events. I modified `dispatch` so it would return a buffer with every event dispatched untouched so we could know when they've been dispatched.

Reply:
I can't stop thinking that a `dispatch()` function extracted from the current main one (which could be renamed to `dispatchLoop()`) would be far easier to `await` in tests.